File: Backend/src/data/entities/Poll.ts
import { Poll_Option } from "./Poll_Option";
import { Vote } from "./Vote";
import { PollCreateDto } from "../../api/polls/models/requests/PollCreate.dto";
import { uuid } from "../../utils/util";
import { ApiProperty } from "@nestjs/swagger";

export enum PollStage {
	NOT_STARTED = "NOT_STARTED",
	IN_PROGRESS = "IN_PROGRESS",
	FINISHED = "FINISHED"
}

export class Poll {
	@ApiProperty()
	id: string; // matches team ID
	@ApiProperty()
	topic: string;
	@ApiProperty()
	description: string;
	@ApiProperty()
	time: Date;
	@ApiProperty()
	location: string;
	@ApiProperty()
	isMultichoice: boolean;
	@ApiProperty({ type: () => [Poll_Option] })
	options: Poll_Option[];
	@ApiProperty({ type: () => [Vote] })
	votes: Vote[];
	@ApiProperty({ enum: PollStage })
	stage: PollStage;

	static fromDto(dto: PollCreateDto, teamId: string): Poll {
		const options: Poll_Option[] = [];
		dto.options.forEach((option) =>
			options.push({
				id: uuid(),
				option: option
			})
		);
		options.push({
			id: "-1",
			option: "Opt Out"
		});
		return {
			id: teamId,
			topic: dto.topic,
			description: dto.description,
			time: dto.time,
			location: dto.location,
			isMultichoice: dto.isMultichoice,
			options: options,
			votes: [],
			stage: PollStage.NOT_STARTED
		};
	}
}

File: frontend/lib/Types/poll.dart
import 'package:frontend/Types/vote.dart';

import 'poll_option.dart';
import 'poll_stage.dart';

// Initializing all the elements of the poll
class Poll {
  String id;
  String topic;
  String description;
  DateTime time;
  String location;
  bool isMultipleChoice;
  List<PollOption> options;
  List<Vote>? votes = [];
  PollStage? stage;

  Poll(this.id, this.topic, this.description, this.time, this.location,
      this.isMultipleChoice, this.options,
      [this.votes, this.stage]);

  factory Poll.fromJson(Map<String, dynamic> json) {
    return Poll(
        json['id'] as String,
        json['topic'] as String,
        json['description'] as String,
        DateTime.parse(json['time'] as String),
        json['location'] as String,
        json['isMultichoice'] as bool,
        json["options"].map<PollOption>((e) => PollOption.fromJson(e)).toList(),
        json["votes"].map<Vote>((e) => Vote.fromJson(e)).toList(),
        PollStage.values
            .firstWhere((e) => e.toString() == 'PollStage.${json["stage"]}'));
  }
}
